File: java/src/main/java/io/github/emindeniz99/applepurchase/jws/AppTransactionPayload.java
package io.github.emindeniz99.applepurchase.jws;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * Decoded {@code AppTransaction} payload (StoreKit 2, iOS 16+): app-level
 * proof of purchase/download. Environment lives in {@link #receiptType()};
 * dates are milliseconds since epoch; {@code null} means the claim was absent.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public final class AppTransactionPayload {

    private Long appAppleId;
    private String appTransactionId;
    private String applicationVersion;
    private String bundleId;
    private String deviceVerification;
    private String deviceVerificationNonce;
    private String originalApplicationVersion;
    private Long originalPurchaseDate;
    private Long preorderDate;
    private Long receiptCreationDate;
    private String receiptType;
    private Long versionExternalIdentifier;

    AppTransactionPayload() {
    }

    public Long appAppleId() {
        return appAppleId;
    }

    public String appTransactionId() {
        return appTransactionId;
    }

    public String applicationVersion() {
        return applicationVersion;
    }

    public String bundleId() {
        return bundleId;
    }

    public String deviceVerification() {
        return deviceVerification;
    }

    public String deviceVerificationNonce() {
        return deviceVerificationNonce;
    }

    public String originalApplicationVersion() {
        return originalApplicationVersion;
    }

    public Long originalPurchaseDate() {
        return originalPurchaseDate;
    }

    public Long preorderDate() {
        return preorderDate;
    }

    public Long receiptCreationDate() {
        return receiptCreationDate;
    }

    /** The environment claim of an AppTransaction (e.g. {@code "Production"}). */
    public String receiptType() {
        return receiptType;
    }

    public Long versionExternalIdentifier() {
        return versionExternalIdentifier;
    }
}

File: dotnet/src/ApplePurchaseReceiptVerifier/Jws/AppTransactionPayload.cs
using System;
using System.Collections.Generic;

namespace ApplePurchaseReceiptVerifier.Jws
{
    /// <summary>
    /// A decoded <c>AppTransaction</c> payload (StoreKit 2, iOS 16+):
    /// app-level proof of purchase or download.
    /// </summary>
    /// <remarks>
    /// The environment lives in <see cref="ReceiptType"/>. Dates are
    /// milliseconds since the epoch, exactly as Apple ships them;
    /// <see langword="null"/> means the claim was absent.
    /// </remarks>
    public sealed class AppTransactionPayload
    {
        internal AppTransactionPayload(IReadOnlyDictionary<string, object?> claims)
        {
            ClaimsMap = claims;
            AppAppleId = Internal.Claims.Int64(claims, "appAppleId");
            AppTransactionId = Internal.Claims.String(claims, "appTransactionId");
            ApplicationVersion = Internal.Claims.String(claims, "applicationVersion");
            BundleId = Internal.Claims.String(claims, "bundleId");
            DeviceVerification = Internal.Claims.String(claims, "deviceVerification");
            DeviceVerificationNonce = Internal.Claims.String(claims, "deviceVerificationNonce");
            OriginalApplicationVersion = Internal.Claims.String(claims, "originalApplicationVersion");
            OriginalPurchaseDate = Internal.Claims.Int64(claims, "originalPurchaseDate");
            PreorderDate = Internal.Claims.Int64(claims, "preorderDate");
            ReceiptCreationDate = Internal.Claims.Int64(claims, "receiptCreationDate");
            ReceiptType = Internal.Claims.String(claims, "receiptType");
            VersionExternalIdentifier = Internal.Claims.Int64(claims, "versionExternalIdentifier");
        }

        /// <summary>Every claim in the verified payload, including unmodelled ones.</summary>
        public IReadOnlyDictionary<string, object?> ClaimsMap { get; }

        /// <summary>The <c>appAppleId</c> claim.</summary>
        public long? AppAppleId { get; }

        /// <summary>The <c>appTransactionId</c> claim.</summary>
        public string? AppTransactionId { get; }

        /// <summary>The <c>applicationVersion</c> claim.</summary>
        public string? ApplicationVersion { get; }

        /// <summary>The <c>bundleId</c> claim.</summary>
        public string? BundleId { get; }

        /// <summary>The <c>deviceVerification</c> claim.</summary>
        public string? DeviceVerification { get; }

        /// <summary>The <c>deviceVerificationNonce</c> claim.</summary>
        public string? DeviceVerificationNonce { get; }

        /// <summary>The <c>originalApplicationVersion</c> claim.</summary>
        public string? OriginalApplicationVersion { get; }

        /// <summary>The <c>originalPurchaseDate</c> claim, in epoch milliseconds.</summary>
        public long? OriginalPurchaseDate { get; }

        /// <summary>The <c>preorderDate</c> claim, in epoch milliseconds.</summary>
        public long? PreorderDate { get; }

        /// <summary>The <c>receiptCreationDate</c> claim, in epoch milliseconds.</summary>
        public long? ReceiptCreationDate { get; }

        /// <summary>The environment claim of an AppTransaction, e.g. <c>"Production"</c>.</summary>
        public string? ReceiptType { get; }

        /// <summary>The <c>versionExternalIdentifier</c> claim.</summary>
        public long? VersionExternalIdentifier { get; }
    }
}
